core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -4882,4 +4947,282 @@ private boolean isH2DB(Connection dbConnection) throws Exception {
+                    " when circuit breaker state is "
+                    + getCircuitBreakerState() + " and circuit breaker open duration: "
+                    + circuitOpenDuration + "ms.");
+        }
+
+        if (circuitOpenDuration >= getThresholdTimeoutInMilliseconds()) {
+            try {
+                Connection dbConnection = getConnection();
+                setCircuitBreakerState(CIRCUIT_STATE_CLOSE);
+                setThresholdStartTime(0);
+                return dbConnection;
+            } catch (Exception e) {
+                log.warn("Error occurred while trying to obtaining connection for domain: " + getMyDomainName()
+                + ". Circuit Breaker state for domain: " + getMyDomainName() + " is set to "
+                        + getCircuitBreakerState());
+                setThresholdStartTime(System.currentTimeMillis());
+
+                throw new CircuitBreakerException("Error occurred while obtaining connection.", e);
+            }
+        } else {
+            throw new CircuitBreakerException(
+                    "JDBC Connection circuit breaker is in open state for " + circuitOpenDuration
+                            + "ms and has not reach the threshold timeout: " + getThresholdTimeoutInMilliseconds()
+                            + "ms, hence avoid establishing the connection for domain " + getMyDomainName());
+        }
+    }
+
+    /**
+     * Attempts and returns the DB Connection when the circuit Breaker is in closed state.
+     * @return returns the DB connection.
+     *
+     * @throws CircuitBreakerException An error occurred while attempting to retrieve the DB connection.
+     */
+    private Connection getConnectionOnCircuitBreakerClose() throws CircuitBreakerException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Connection circuit breaker state: " + getCircuitBreakerState()
+                    + ", so trying to obtain the connection for domain " + getMyDomainName());
+        }
+        int retryCounter = 0;
+        while (true) {
+            try {
+                return getConnection();
+            } catch (Exception e) {
+                log.warn("Error occurred while obtaining JDBC connection for domain " + getMyDomainName()
+                + ". Hence, retry attempt to recover DB connection: " + retryCounter);
+
+                if (++retryCounter >= getConnectionRetryCount()) {
+                    log.error("Retry count exceeds above the maximum count: " + getConnectionRetryCount() +
+                            " and failed for domain " + getMyDomainName());
+                    setCircuitBreakerState(CIRCUIT_STATE_OPEN);
+                    setThresholdStartTime(System.currentTimeMillis());
+                    throw new CircuitBreakerException("Error occurred while obtaining connection for domain: "
+                            + getMyDomainName() + " and circuit breaker state set to: " + getCircuitBreakerState(), e);
+                }
```

Comment:
isn't it better, if we wait here a bit rather than trying to obtain the connection continuously?

Won't this overload the DB instance which is already in stress

Reply:
I agree with your perspective, there is a possibility for such an event to occur. Do you think specifying a default wait time would be ideal in this case?

Reply:
what will be the default connection retry count?

Reply:
the default connection retry count I added is 2 times.